source/common/upstream/odcds_api_impl.cc

```diff
@@ -0,0 +1,85 @@
+#include "common/upstream/odcds_api_impl.h"
```

Comment:
Sorry, I'm a bit late to the party here, but why do we have a different CDS implementation for on-demand vs. regular CDS? In all the other on-demand models (VHDS, SRDS), we reuse the existing machinery and just making it on-demand aware. Here we have diverged CDS implementation. This will probably have implications down the track when we make bug fixes for regular delta CDS but miss out on on OCDS and vice versa.

I would like to see better awareness of on-demand as a horizontal in Envoy that spans across subsystems, rather than doing what is expedient or local to CDS for OCDS here. 

CC @mattklein123 @adisuissa @dmitri-d

Reply:
> CC @mattklein123 @adisuissa @dmitri-d

As I said offline, I haven't actually reviewed the code in detail yet, just the interfaces. I will review in more detail today. My expectation is most/all of the code is shared. I'm not sure it is possible to implement on-demand in main CDS due to the SoTW wildcard semantics, but we can see (at a certain point though it may just mean moving the same code split to a different location).

Reply:
> Sorry, I'm a bit late to the party here, but why do we have a different CDS implementation for on-demand vs. regular CDS?

There is not much of the code difference really - most of the code is shared through the CdsApiHelper. The differences between CDS and ODCDS are mostly about when to start subscription (for CDS is sometime at Envoy startup, for ODCDS is on the first request). I suppose I could make CDS on-demand aware, but I'm afraid that would result in awkward code inside it.

> In all the other on-demand models (VHDS, SRDS), we reuse the existing machinery and just making it on-demand aware. Here we have diverged CDS implementation. This will probably have implications down the track when we make bug fixes for regular delta CDS but miss out on on OCDS and vice versa.

I tried to move most of the code into the helper class so such bug fixes would most certainly land in a helper class, not in CDS/ODCDS itself.

> 
> I would like to see better awareness of on-demand as a horizontal in Envoy that spans across subsystems, rather than doing what is expedient or local to CDS for OCDS here.
> 
> CC @mattklein123 @adisuissa @dmitri-d